tough-kms/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+//! tough-kms implements the `KeySource` trait found in [tough, a Rust TUF client](https://github.com/awslabs/tough).
+//!
+//! By implementing this trait, AWS KMS can become a source of keys used to sign a [TUF repository](https://theupdateframework.github.io/).
+//!
+//! # Testing
+//!
+//! Unit tests are run in the usual manner: `cargo test`.
+
+#![forbid(missing_debug_implementations, missing_copy_implementations)]
+#![deny(rust_2018_idioms)]
+// missing_docs is on its own line to make it easy to comment out when making changes.
+#![deny(missing_docs)]
+#![warn(clippy::pedantic)]
+#![allow(
+    clippy::module_name_repetitions,
+    clippy::must_use_candidate,
+    clippy::missing_errors_doc
+)]
+
+mod client;
+pub mod error;
+use ring::rand::SecureRandom;
+use rusoto_kms::{Kms, KmsClient, SignRequest};
+use snafu::{OptionExt, ResultExt};
+use std::collections::HashMap;
+use std::fmt;
+use tough::key_source::KeySource;
+use tough::schema::decoded::{Decoded, RsaPem};
+use tough::schema::key::{Key, RsaKey, RsaScheme};
+use tough::sign::Sign;
+
+/// Represents a Signing Algorithms for AWS KMS.
+#[derive(Debug, Clone, PartialEq)]
+pub enum KmsSigningAlgorithms {
+    /// The key type
+    Rsa(String),
+}
```

Comment:
I'm confused about why we need this type. What advantage does this have over `KmsRsaKey.signing_algorithm` just being a `String`?

Reply:
Yes, It is confusing, I would like to have your feedback on this.  We had a discussion on question as described below
>  KMS supports a lot of signing methods

> > Iliana Reply: What I'd recommend is a match statement, even though the enum has one field. That way if we add new ones because TUF allows new ones, it's more clear that we need to change downstream code.

It was added in context of this, to keep it generic and make decision of signing algorithm in  `as_sign` while returning the `sign` object:
```
Ok(Box::new(KmsRsaKey {
            profile: self.profile.clone(),
            client: Some(kms_client.clone()),
            key_id: self.key_id.clone(),
            public_key: pb_key,
            signing_algorithm: KmsSigningAlgorithms::Rsa(String::from("RSASSA_PSS_SHA_256")),
        }))
```

However, It does not solve the problem in a way I wanted to.  I want something like object of algorithms to select from.
Ex:-
```
algorithms = {
   algo1 : "algo1",
   algo2: "algo2"
} 
 And we can directly use as `algorithms.algo1` or ``algorithms.algo2` based on some conditions in `as_sign` in future when we support more signing algorithms.

Reply:
OK, I think I see what's necessary here. We need to support a way of saying "use this particular signing algorithm with KMS" in case the TUF spec adds additional signing algorithms for the same kind of key.

If that's the case, it needs to be a member of `KmsKeySource` so that the user can specify a different algorithm. And I would avoid using strings at all in this circumstance.

Here's what I'd propose:

```rust
#[non_exhaustive]
pub enum KmsSigningAlgorithm {
    RsassaPssSha256,
}

pub struct KmsKeySource {
    ...
    pub signing_algorithm: KmsSigningAlgorithm,
}
```

Then, verify the GetPublicKey response contains a value that matches `signing_algorithm`, returning an error if there's a mismatch (perhaps the key in KMS is ECDSA).

Because there's only one option implemented for now, I don't think we need to worry about wiring this up all the way through tuftool yet.

Reply:
Done.